File: backend/src/test/java/com/coing/domain/coin/orderbook/service/OrderbookServiceTest.java
package com.coing.domain.coin.orderbook.service;

import static org.junit.jupiter.api.Assertions.*;
import static org.mockito.Mockito.*;

import java.util.LinkedList;

import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.DisplayName;
import org.junit.jupiter.api.Test;
import org.junit.jupiter.api.extension.ExtendWith;
import org.mockito.ArgumentCaptor;
import org.mockito.InjectMocks;
import org.mockito.Mock;
import org.mockito.junit.jupiter.MockitoExtension;
import org.springframework.messaging.simp.SimpMessageSendingOperations;

import com.coing.domain.coin.orderbook.dto.OrderbookDto;
import com.coing.domain.coin.orderbook.entity.Orderbook;
import com.coing.domain.coin.orderbook.entity.OrderbookUnit;
import com.fasterxml.jackson.core.JsonProcessingException;
import com.fasterxml.jackson.databind.ObjectMapper;

@ExtendWith(MockitoExtension.class)
public class OrderbookServiceTest {
	@Mock
	SimpMessageSendingOperations simpMessageSendingOperations;

	@InjectMocks
	OrderbookService orderbookService;

	private Orderbook testOrderbook;
	private final ObjectMapper mapper = new ObjectMapper();

	@BeforeEach
	public void setUp() {
		OrderbookUnit unit = new OrderbookUnit(100.0, 90.0, 10.0, 5.0);
		LinkedList<OrderbookUnit> units = new LinkedList<>();
		units.add(unit);

		testOrderbook = Orderbook.builder()
			.type("orderbook")
			.code("KRW-BTC")
			.totalAskSize(10.0)
			.totalBidSize(5.0)
			.orderbookUnits(units)
			.timestamp(System.currentTimeMillis())
			.level(0.0)
			.build();
	}

	@Test
	@DisplayName("publish 성공")
	void publish() throws JsonProcessingException {
		// given

		OrderbookDto dto = OrderbookDto.from(testOrderbook);
		// when
		orderbookService.publish(dto);

		// then
		String expectedChannel = "/sub/coin/orderbook/" + testOrderbook.getCode();
		ArgumentCaptor<OrderbookDto> captor = ArgumentCaptor.forClass(OrderbookDto.class);

		verify(simpMessageSendingOperations, times(1))
			.convertAndSend(eq(expectedChannel), captor.capture());

		OrderbookDto sentDto = captor.getValue();
		assertEquals("orderbook", sentDto.type());
		assertEquals("KRW-BTC", sentDto.code());
	}
}

File: backend/src/test/java/com/coing/domain/coin/orderbook/service/OrderbookServiceTest.kt
package com.coing.domain.coin.orderbook.service

import com.coing.domain.coin.common.port.EventPublisher
import com.coing.domain.coin.orderbook.dto.OrderbookDto
import com.coing.domain.coin.orderbook.entity.Orderbook
import com.coing.domain.coin.orderbook.entity.OrderbookUnit
import org.junit.jupiter.api.Assertions.assertEquals
import org.junit.jupiter.api.BeforeEach
import org.junit.jupiter.api.DisplayName
import org.junit.jupiter.api.Test
import org.mockito.kotlin.*
import kotlin.test.assertTrue

class OrderbookServiceTest {
    private val orderbookPublisher: EventPublisher<OrderbookDto> = mock()
    private lateinit var orderbookService: OrderbookService
    private lateinit var testOrderbook: Orderbook
    private lateinit var testOrderbook2: Orderbook

    @BeforeEach
    fun setUp() {
        orderbookService = OrderbookService(orderbookPublisher)
        val unit = OrderbookUnit(
            askPrice = 100.0,
            bidPrice = 90.0,
            askSize = 10.0,
            bidSize = 5.0
        )

        testOrderbook = Orderbook(
            type = "orderbook",
            code = "KRW-BTC",
            totalAskSize = 10.0,
            totalBidSize = 5.0,
            orderbookUnits = listOf(unit),
            timestamp = System.currentTimeMillis(),
            level = 0.0
        )

        testOrderbook2 = Orderbook(
            type = "orderbook",
            code = "KRW-ETH",
            totalAskSize = 10.0,
            totalBidSize = 5.0,
            orderbookUnits = listOf(unit),
            timestamp = System.currentTimeMillis(),
            level = 0.0
        )
    }

    @Test
    @DisplayName("update 성공")
    fun update() {
        // given
        val captor = argumentCaptor<OrderbookDto>()

        // when
        orderbookService.update(testOrderbook)

        // then
        verify(orderbookPublisher, times(1))
            .publish(captor.capture())
        assertEquals("KRW-BTC", captor.firstValue.code)
    }

    @Test
    @DisplayName("publish 성공")
    fun publish() {
        // given
        val dto = OrderbookDto.from(testOrderbook)

        // when
        orderbookService.publish(dto)

        // then
        val expectedChannel = "/sub/coin/orderbook/${testOrderbook.code}"
        argumentCaptor<OrderbookDto>().apply {
            verify(orderbookPublisher, times(1)).publish(capture())
            val sentDto = firstValue

            assertEquals("orderbook", sentDto.type)
            assertEquals("KRW-BTC", sentDto.code)
        }
    }

    @Test
    @DisplayName("fallbackUpdate 성공")
    fun fallbackUpdate() {
        // given
        orderbookService.update(testOrderbook)
        orderbookService.update(testOrderbook2)

        // when
        orderbookService.fallbackUpdate("13:05:22")

        // then
        val cachedData = orderbookService.getAllCachedData()
        cachedData.forEach {
            assertTrue(it.isFallback)
            assertEquals("13:05:22", it.lastUpdate)
        }

        verify(orderbookPublisher, times(2)).publish(argThat {
            isFallback && lastUpdate == "13:05:22"
        })
    }
}
